// skill-engine/config.js
'use strict';

const { SKILL_MAP } = require('skill-model');

const DEFAULT_RESOURCE_TYPE = 'qi';

const BASIC_ATTACK = Object.freeze({
  id: 'basic_attack',
  name: '普攻',
  type: 'basic',
  cooldown: 0,
  animation: 'basic',
  damage: [
    {
      kind: 'auto'
    }
  ],
  resource: { type: DEFAULT_RESOURCE_TYPE, cost: 0 }
});

const SKILL_RUNTIME_LIBRARY = Object.freeze({
  sword_breaking_clouds: {
    id: 'sword_breaking_clouds',
    name: '破云斩',
    type: 'active',
    cooldown: 2,
    animation: 'sword_slash',
    damage: [
      {
        kind: 'physical',
        ratio: 1.2,
        perLevel: 0.04,
        extraOnCritRatio: 0.3,
        extraOnCritPerLevel: 0.02
      }
    ]
  },
  body_blood_fury: {
    id: 'body_blood_fury',
    name: '熔血怒元',
    type: 'active',
    cooldown: 4,
    animation: 'body_enrage',
    healSelf: {
      ratioOfMaxHp: 0.15,
      perLevel: 0.008
    },
    buffs: [
      {
        target: 'self',
        type: 'reflect',
        ratio: 0.1,
        perLevel: 0.01,
        duration: 2
      }
    ]
  },
  sword_flowing_strike: {
    id: 'sword_flowing_strike',
    name: '流光剑步',
    type: 'active',
    cooldown: 4,
    animation: 'dash_strike',
    damage: [
      {
        kind: 'physical',
        ratio: 1.4,
        perLevel: 0.03
      }
    ],
    buffs: [
      {
        target: 'self',
        type: 'stat',
        stat: 'speed',
        mode: 'percent',
        value: 0.12,
        perLevel: 0.005,
        duration: 2
      }
    ]
  },
  spell_burning_burst: {
    id: 'spell_burning_burst',
    name: '烈炽火弹',
    type: 'active',
    cooldown: 3,
    animation: 'fireball',
    damage: [
      {
        kind: 'magic',
        ratio: 1,
        perLevel: 0.02
      }
    ],
    dots: [
      {
        type: 'burning',
        attribute: 'magicAttack',
        ratio: 0.2,
        perLevel: 0.02,
        duration: 2
      }
    ]
  },
  spell_frost_bolt: {
    id: 'spell_frost_bolt',
    name: '凝霜矢',
    type: 'active',
    cooldown: 3,
    animation: 'ice_bolt',
    damage: [
      {
        kind: 'magic',
        ratio: 1.3,
        perLevel: 0.03
      }
    ],
    debuffs: [
      {
        target: 'opponent',
        type: 'stat',
        stat: 'speed',
        mode: 'percent',
        value: -0.3,
        perLevel: -0.01,
        duration: 2
      }
    ]
  },
  body_blood_ignite: {
    id: 'body_blood_ignite',
    name: '焚血激',
    type: 'active',
    cooldown: 4,
    animation: 'body_enrage',
    healSelf: {
      ratioOfMaxHp: 0.15,
      perLevel: 0.01
    },
    buffs: [
      {
        target: 'self',
        type: 'reflect',
        ratio: 0.1,
        perLevel: 0.01,
        duration: 2
      }
    ]
  },
  beast_spirit_pact: {
    id: 'beast_spirit_pact',
    name: '灵契术',
    type: 'active',
    cooldown: 4,
    animation: 'generic',
    buffs: [
      {
        target: 'self',
        type: 'stat',
        stat: 'summonPower',
        mode: 'percent',
        value: 0.15,
        perLevel: 0.01,
        duration: 3
      }
    ]
  },
  beast_wood_blessing: {
    id: 'beast_wood_blessing',
    name: '木灵惠泽',
    type: 'active',
    cooldown: 5,
    animation: 'generic',
    healSelf: {
      ratioOfMaxHp: 0.18,
      perLevel: 0.008
    },
    buffs: [
      {
        target: 'self',
        type: 'stat',
        stat: 'healingBonus',
        mode: 'percent',
        value: 0.2,
        perLevel: 0.01,
        duration: 3
      }
    ]
  },
  spell_frost_tide: {
    id: 'spell_frost_tide',
    name: '凌霜定潮',
    type: 'active',
    cooldown: 4,
    animation: 'ice_wave',
    damage: [
      {
        kind: 'magic',
        ratio: 1.3,
        perLevel: 0.02
      }
    ],
    debuffs: [
      {
        target: 'opponent',
        type: 'stat',
        stat: 'finalDamageBonus',
        mode: 'flat',
        value: -0.2,
        perLevel: -0.01,
        duration: 2
      }
    ]
  },
  spell_frost_prison: {
    id: 'spell_frost_prison',
    name: '霜渊天缚',
    type: 'active',
    cooldown: 5,
    animation: 'ice_prison',
    damage: [
      {
        kind: 'magic',
        ratio: 1.7,
        perLevel: 0.04
      }
    ],
    control: {
      type: 'stun',
      chance: 0.6,
      perLevel: 0.02,
      duration: 2
    },
    followUp: {
      type: 'burst',
      ratio: 0.4,
      kind: 'magic'
    }
  },
  spell_searing_comet: {
    id: 'spell_searing_comet',
    name: '灼脉流炬',
    type: 'active',
    cooldown: 4,
    animation: 'fire_lane',
    damage: [
      {
        kind: 'magic',
        ratio: 1.45,
        perLevel: 0.03
      }
    ],
    conditionalDamage: [
      {
        condition: 'burning',
        kind: 'magic',
        ratio: 0.3,
        perLevel: 0.02
      }
    ]
  },
  sword_flame_wings: {
    id: 'sword_flame_wings',
    name: '烈羽焚锋',
    type: 'active',
    cooldown: 4,
    animation: 'fire_slash',
    damage: [
      {
        kind: 'physical',
        ratio: 1.35,
        perLevel: 0.03,
        hits: 2
      }
    ],
    dots: [
      {
        type: 'burning',
        attribute: 'physicalAttack',
        ratio: 0.15,
        perLevel: 0.01,
        duration: 2
      }
    ]
  },
  sword_thunder_break: {
    id: 'sword_thunder_break',
    name: '雷霆断界',
    type: 'active',
    cooldown: 4,
    animation: 'thunder_strike',
    damage: [
      {
        kind: 'physical',
        ratio: 0.6,
        perLevel: 0.04,
        hits: 3
      }
    ],
    control: {
      type: 'stun',
      chance: 0.5,
      perLevel: 0.02,
      duration: 1
    }
  },
  spell_thunder_chain: {
    id: 'spell_thunder_chain',
    name: '万雷劫链',
    type: 'active',
    cooldown: 6,
    animation: 'lightning_chain',
    damage: [
      {
        kind: 'magic',
        ratio: 2.2,
        perLevel: 0.05
      }
    ],
    control: {
      type: 'stun',
      chance: 0.4,
      perLevel: 0.03,
      duration: 1
    }
  },
  sword_thousand_blades: {
    id: 'sword_thousand_blades',
    name: '千刃星陨',
    type: 'active',
    cooldown: 4,
    animation: 'blade_storm',
    damage: [
      {
        kind: 'physical',
        ratio: 0.7,
        perLevel: 0.03,
        hits: 4
      }
    ]
  },
  sigil_corroding_mark: {
    id: 'sigil_corroding_mark',
    name: '蚀骨符',
    type: 'active',
    cooldown: 4,
    animation: 'poison_mark',
    debuffs: [
      {
        target: 'opponent',
        type: 'stat',
        stat: 'physicalDefense',
        mode: 'percent',
        value: -0.1,
        perLevel: -0.005,
        duration: 3
      },
      {
        target: 'opponent',
        type: 'stat',
        stat: 'magicDefense',
        mode: 'percent',
        value: -0.1,
        perLevel: -0.005,
        duration: 3
      }
    ],
    dots: [
      {
        type: 'poison',
        attribute: 'magicAttack',
        ratio: 0.1,
        perLevel: 0.005,
        duration: 3
      }
    ]
  },
  sigil_focus_talisman: {
    id: 'sigil_focus_talisman',
    name: '定神符',
    type: 'active',
    cooldown: 5,
    animation: 'sigil_shock',
    damage: [
      {
        kind: 'magic',
        ratio: 0.5
      }
    ],
    control: {
      type: 'stun',
      chance: 0.8,
      perLevel: 0.02,
      duration: 1
    }
  },
  sigil_soul_bind: {
    id: 'sigil_soul_bind',
    name: '镇魂神符',
    type: 'active',
    cooldown: 5,
    animation: 'sigil_bind',
    control: {
      type: 'silence',
      chance: 0.7,
      perLevel: 0.02,
      duration: 2
    },
    debuffs: [
      {
        target: 'opponent',
        type: 'stat',
        stat: 'healingReceived',
        mode: 'flat',
        value: -0.2,
        perLevel: -0.02,
        duration: 2
      }
    ]
  },
  sigil_rupture_chain: {
    id: 'sigil_rupture_chain',
    name: '断厄符索',
    type: 'active',
    cooldown: 4,
    animation: 'sigil_chain',
    damage: [
      {
        kind: 'magic',
        ratio: 1.2,
        perLevel: 0.02
      }
    ],
    debuffs: [
      {
        target: 'opponent',
        type: 'stat',
        stat: 'rageGain',
        mode: 'flat',
        value: -0.2,
        perLevel: -0.01,
        duration: 3
      }
    ]
  },
  sword_immortal_domain: {
    id: 'sword_immortal_domain',
    name: '戮仙剑域',
    type: 'active',
    cooldown: 6,
    animation: 'sword_domain',
    buffs: [
      {
        target: 'self',
        type: 'extraDamage',
        ratio: 0.5,
        perLevel: 0.02,
        duration: 3
      }
    ]
  },
  spell_pyrocataclysm: {
    id: 'spell_pyrocataclysm',
    name: '离火焚天',
    type: 'active',
    cooldown: 5,
    animation: 'firestorm',
    damage: [
      {
        kind: 'magic',
        ratio: 1.8,
        perLevel: 0.04
      }
    ],
    debuffs: [
      {
        target: 'opponent',
        type: 'stat',
        stat: 'finalDamageReduction',
        mode: 'flat',
        value: -0.15,
        perLevel: -0.01,
        duration: 3
      }
    ],
    dots: [
      {
        type: 'burning',
        attribute: 'magicAttack',
        ratio: 0.3,
        perLevel: 0.015,
        duration: 3
      }
    ]
  },
  body_furnace_of_ruin: {
    id: 'body_furnace_of_ruin',
    name: '焚世熔炉',
    type: 'active',
    cooldown: 6,
    animation: 'flame_barrier',
    buffs: [
      {
        target: 'self',
        type: 'shield',
        ratioOfMaxHp: 0.4,
        perLevel: 0.03,
        duration: 3,
        reflectRatio: 0.2,
        reflectPerLevel: 0.01,
        burstRatio: 2.5
      }
    ]
  },
  sigil_taiyi_barrier: {
    id: 'sigil_taiyi_barrier',
    name: '太乙护界',
    type: 'active',
    cooldown: 7,
    animation: 'earth_barrier',
    buffs: [
      {
        target: 'self',
        type: 'shield',
        ratioOfMaxHp: 0.3,
        perLevel: 0.02,
        duration: 3
      },
      {
        target: 'self',
        type: 'stat',
        stat: 'finalDamageReduction',
        mode: 'flat',
        value: 0.2,
        perLevel: 0.01,
        duration: 3
      }
    ]
  },
  beast_empyrial_charge: {
    id: 'beast_empyrial_charge',
    name: '帝御九霄',
    type: 'active',
    cooldown: 6,
    animation: 'generic',
    buffs: [
      {
        target: 'self',
        type: 'stat',
        stat: 'summonPower',
        mode: 'percent',
        value: 0.4,
        perLevel: 0.02,
        duration: 2
      },
      {
        target: 'self',
        type: 'extraDamage',
        ratio: 1,
        perLevel: 0.05,
        duration: 1
      }
    ]
  }
});

function resolveSkillCost(skillId) {
  if (!skillId) {
    return 0;
  }
  const definition = SKILL_MAP[skillId];
  if (definition && definition.params && Number.isFinite(definition.params.cost)) {
    return Math.max(0, Math.round(definition.params.cost));
  }
  return 0;
}

function withResourceDefinition(runtime, skillId) {
  const cost = resolveSkillCost(skillId);
  const resourceType = (runtime.resource && runtime.resource.type) || DEFAULT_RESOURCE_TYPE;
  return {
    ...runtime,
    resource: {
      type: resourceType,
      cost
    }
  };
}

function resolveRuntimeSkill(skillId) {
  if (!skillId) {
    return withResourceDefinition({ ...BASIC_ATTACK }, BASIC_ATTACK.id);
  }
  const definition = SKILL_RUNTIME_LIBRARY[skillId];
  if (definition) {
    return withResourceDefinition({ ...definition }, skillId);
  }
  const fallback = SKILL_MAP[skillId];
  if (fallback) {
    const runtime = {
      id: fallback.id,
      name: fallback.name,
      type: fallback.type || 'active',
      cooldown: fallback.params && Number.isFinite(fallback.params.cooldown)
        ? Math.max(1, Math.floor(fallback.params.cooldown))
        : 3,
      animation: 'generic',
      damage: [
        {
          kind: fallback.discipline === 'spell' ? 'magic' : 'physical',
          ratio: 1.1
        }
      ]
    };
    return withResourceDefinition(runtime, skillId);
  }
  return withResourceDefinition({ ...BASIC_ATTACK }, BASIC_ATTACK.id);
}

module.exports = {
  BASIC_ATTACK,
  SKILL_RUNTIME_LIBRARY,
  resolveRuntimeSkill
};
