// common-config/index.js
const subLevelLabels = ['一层', '二层', '三层', '四层', '五层', '六层', '七层', '八层', '九层', '圆满'];

const EXPERIENCE_PER_YUAN = 100;

const COLLECTIONS = Object.freeze({
  MEMBERS: 'members',
  LEVELS: 'membershipLevels',
  MEMBERSHIP_LEVELS: 'membershipLevels',
  RIGHTS_MASTER: 'membershipRights',
  MEMBERSHIP_RIGHTS: 'membershipRights',
  MEMBER_RIGHTS: 'memberRights',
  MEMBER_EXTRAS: 'memberExtras',
  MEMBER_TIMELINE: 'memberTimeline',
  MENU_ORDERS: 'menuOrders',
  CHARGE_ORDERS: 'chargeOrders',
  RESERVATIONS: 'reservations',
  ROOMS: 'rooms',
  TASKS: 'tasks',
  TASK_RECORDS: 'taskRecords',
  COUPONS: 'coupons',
  COUPON_RECORDS: 'couponRecords',
  STONE_TRANSACTIONS: 'stoneTransactions',
  WALLET_TRANSACTIONS: 'walletTransactions',
  TRANSACTIONS: 'walletTransactions',
  AVATAR_CATEGORIES: 'avatarCategories',
  AVATARS: 'avatars',
  PVP_PROFILES: 'pvpProfiles',
  PVP_SEASONS: 'pvpSeasons',
  PVP_MATCHES: 'pvpMatches',
  PVP_LEADERBOARD: 'pvpLeaderboard',
  PVP_INVITES: 'pvpInvites',
  ERROR_LOGS: 'errorlogs'
});

const EXCLUDED_TRANSACTION_STATUSES = Object.freeze([
  'pending',
  'processing',
  'failed',
  'cancelled',
  'refunded',
  'closed'
]);

const DEFAULT_ADMIN_ROLES = Object.freeze(['admin', 'developer']);

const realmConfigs = [
  {
    id: 'realm_qi_refining',
    name: '炼气期',
    shortName: '炼气',
    description: '入门境界，通过吸纳天地灵气淬炼肉身，正式踏上修行大道。',
    increment: 100,
    thresholds: [0, 200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800],
    discount: 1,
    virtualRewards: [
      '称号「炼气·初悟」，解锁灵根检测背景',
      '灵石积分 +100，可兑换入门灵草',
      '头像框「灵气萌芽」开放',
      '修炼服饰「青布道袍」上线',
      '虚拟法器「灵木剑」挂件',
      '聊天气泡「灵气环绕」',
      '消费积分加成 +5%，基础任务收益提升',
      '表情包「闭关修炼」解锁',
      '灵宠挂件「碧灵狐」幼崽随行',
      '修炼笔记皮肤「炼气圆满」开放'
    ],
    milestone: {
      type: '虚拟+饮品礼遇',
      summary: '灵泉饮品券 1 张 + 限定背景「炼气圆满」',
      rights: [
        {
          rightId: 'right_realm_qi_drink',
          quantity: 1,
          description: '灵泉饮品券 1 张'
        }
      ]
    }
  },
  {
    id: 'realm_foundation',
    name: '筑基期',
    shortName: '筑基',
    description: '稳固根基，凝练真元，寿元翻倍，为结丹做足准备。',
    increment: 200,
    thresholds: [2000, 2200, 2400, 2600, 2800, 3000, 3200, 3400, 3600, 3800],
    discount: 0.99,
    virtualRewards: [
      '称号「筑基·初成」，昵称前缀升级；累计充值 ¥2,000 解锁筑基一层，赠筑基背景与 ¥200 内商品任选 1 件',
      '灵石积分 +200，可兑换筑基丹材料',
      '头像框「筑基丹炉」解锁',
      '服饰「玄铁束甲」上线',
      '本命法器「赤焰飞剑」虚拟道具',
      '聊天气泡「丹火流光」特效',
      '消费积分加成提升至 +10%',
      '表情动作「丹炉炼制」',
      '灵宠进化为「碧灵狐·少成」',
      '专属动作「丹火腾空」展示'
    ],
    milestone: {
      type: '包房体验',
      summary: '私人包房券 1 次（1 小时）+ 虚拟特效「筑基雷纹」',
      rights: [
        {
          rightId: 'right_realm_foundation_room',
          quantity: 1,
          description: '筑基专属包房券 1 次'
        }
      ]
    }
  },
  {
    id: 'realm_core',
    name: '金丹期',
    shortName: '金丹',
    description: '凝聚金丹，可御空飞行，是门派的核心力量。',
    increment: 400,
    thresholds: [5000, 5500, 6000, 6500, 7000, 7500, 8000, 8500, 9000, 9500],
    discount: 0.98,
    virtualRewards: [
      '称号「金丹·初凝」，获得金丹铭牌；累计充值 ¥5,000 解锁金丹一层，赠金丹背景与包房使用次数 1 次',
      '灵石积分 +400，可兑换珍稀材料',
      '头像框「金丹护体」解锁',
      '服饰「锦云法袍」上线',
      '法宝「玉衡飞剑」虚拟模型',
      '聊天气泡「丹光环绕」',
      '消费积分加成提升至 +15%',
      '动作「金丹出窍」解锁',
      '灵宠化形为「碧灵狐·灵巧」',
      '专属座驾「金丹祥云」虚拟坐骑'
    ],
    milestone: {
      type: '定制礼盒',
      summary: '定制礼盒「金丹入怀」 + 稀有法宝皮肤',
      rights: [
        {
          rightId: 'right_realm_core_gift',
          quantity: 1,
          description: '金丹定制礼盒 1 份'
        }
      ]
    }
  },
  {
    id: 'realm_nascent',
    name: '元婴期',
    shortName: '元婴',
    description: '金丹化婴，元神可离体而存，寿元再度提升。',
    increment: 800,
    thresholds: [10000, 11000, 12000, 13000, 14000, 15000, 16000, 17000, 18000, 19000],
    discount: 0.97,
    virtualRewards: [
      '称号「元婴·初现」，展示元神光影；累计充值 ¥10,000 解锁元婴一层，赠元婴背景与包房使用次数 1 次',
      '灵石积分 +800，用于兑换元婴培元丹',
      '头像框「元婴护持」解锁',
      '服饰「素银道袍」上线',
      '法宝「紫雷珠」虚拟特效',
      '聊天气泡「元神出窍」',
      '消费积分加成提升至 +20%',
      '动作「元婴抱元」解锁',
      '灵宠成长为「碧灵狐·灵婴」',
      '开启随身元婴分身虚拟宠物'
    ],
    milestone: {
      type: '精品雪茄',
      summary: '精品雪茄「云雾上品」 + 元婴分身主题',
      rights: [
        {
          rightId: 'right_realm_nascent_cigar',
          quantity: 1,
          description: '元婴精品雪茄 1 支'
        }
      ]
    }
  },
  {
    id: 'realm_divine',
    name: '化神期',
    shortName: '化神',
    description: '感应法则之力，法力不绝，立于人界巅峰门槛。',
    increment: 1600,
    thresholds: [20000, 22000, 24000, 26000, 28000, 30000, 32000, 34000, 36000, 38000],
    discount: 0.96,
    virtualRewards: [
      '称号「化神·初悟」，获得法则光纹；累计充值 ¥20,000 解锁化神一层，赠化神背景（权益待定，可随时调整）',
      '灵石积分 +1600，可兑换化神法则碎片',
      '头像框「化神神光」解锁',
      '服饰「星辉法衣」上线',
      '法宝「九霄神剑」虚拟特效',
      '聊天气泡「神识寰宇」',
      '消费积分加成提升至 +25%',
      '动作「法则降临」解锁',
      '灵宠晋升为「碧灵狐·化神」',
      '解锁飞升试炼场主题'
    ],
    milestone: {
      type: '珍藏红酒',
      summary: '珍藏红酒「神霖年份」 + 化神法则光翼',
      rights: [
        {
          rightId: 'right_realm_divine_wine',
          quantity: 1,
          description: '化神珍藏红酒 1 瓶'
        }
      ]
    }
  },
  {
    id: 'realm_void',
    name: '炼虚期',
    shortName: '炼虚',
    description: '元婴化虚，可凝练分身，调动更广阔的天地之力。',
    increment: 3200,
    thresholds: [50000, 54000, 58000, 62000, 66000, 70000, 74000, 78000, 82000, 86000],
    discount: 0.95,
    virtualRewards: [
      '称号「炼虚·洞明」，附带虚化光影；累计充值 ¥50,000 解锁炼虚一层，赠炼虚背景（权益待定，可随时调整）',
      '灵石积分 +3200，用于兑换虚空晶石',
      '头像框「虚空流光」解锁',
      '服饰「雾隐斗篷」上线',
      '法宝「虚影分身」特效挂件',
      '聊天气泡「虚实交汇」',
      '消费积分加成提升至 +30%',
      '动作「虚空步」解锁',
      '灵宠化虚为「碧灵狐·虚灵」',
      '获得尊贵徽记「炼虚印记」'
    ],
    milestone: {
      type: '尊贵标识',
      summary: '尊贵标识礼包 + 虚空行走特效',
      rights: [
        {
          rightId: 'right_realm_void_badge',
          quantity: 1,
          description: '炼虚尊贵标识礼包'
        }
      ]
    }
  },
  {
    id: 'realm_unity',
    name: '合体期',
    shortName: '合体',
    description: '元神与分身合一，显化法相真身，掌控磅礴灵力。',
    increment: 6400,
    thresholds: [100000, 108000, 116000, 124000, 132000, 140000, 148000, 156000, 164000, 172000],
    discount: 0.94,
    virtualRewards: [
      '称号「合体·初合」，显化合体光环；累计充值 ¥100,000 解锁合体一层，赠合体背景（权益待定，可随时调整）',
      '灵石积分 +6400，兑换合体灵材',
      '头像框「合体法相」解锁',
      '服饰「鸿蒙战铠」上线',
      '法宝「法相金身」虚拟模型',
      '聊天气泡「天地同调」',
      '消费积分加成提升至 +35%',
      '动作「法相显圣」解锁',
      '灵宠晋阶为「碧灵狐·法相」',
      '专属动态光环「合体真身」'
    ],
    milestone: {
      type: '钻石形象与尊享服务',
      summary: '专属虚拟形象套装 + 钻石级尊享服务礼遇',
      rights: [
        {
          rightId: 'right_realm_unity_avatar',
          quantity: 1,
          description: '合体专属形象套装'
        }
      ]
    }
  },
  {
    id: 'realm_great_vehicle',
    name: '大乘期',
    shortName: '大乘',
    description: '修为圆满，掌控更高层次法则之力，距离飞升只差临门一脚。',
    increment: 12800,
    thresholds: [200000, 212000, 224000, 236000, 248000, 260000, 272000, 284000, 296000, 308000],
    discount: 0.93,
    virtualRewards: [
      '称号「大乘·初悟」，获得道韵光芒；累计充值 ¥200,000 解锁大乘一层，赠大乘背景（权益待定，可随时调整）',
      '灵石积分 +12800，可兑换飞升宝材',
      '头像框「大乘法象」解锁',
      '服饰「苍穹霞衣」上线',
      '法宝「鸿蒙界扇」虚拟特效',
      '聊天气泡「道音回响」',
      '消费积分加成提升至 +40%',
      '动作「道韵加身」解锁',
      '灵宠化为「碧灵狐·玄极」',
      '专属场景「大乘天宫」背景'
    ],
    milestone: {
      type: '贵宾宴请',
      summary: '贵宾宴请券 + 大乘天宫场景永久解锁',
      rights: [
        {
          rightId: 'right_realm_great_vehicle_banquet',
          quantity: 1,
          description: '大乘贵宾宴请券'
        }
      ]
    }
  },
  {
    id: 'realm_tribulation',
    name: '渡劫期',
    shortName: '渡劫',
    description: '迎接雷劫洗礼，冲击仙界门槛，成败在此一举。',
    increment: 25600,
    thresholds: [500000, 520000, 540000, 560000, 580000, 600000, 620000, 640000, 660000, 680000],
    discount: 0.92,
    virtualRewards: [
      '称号「渡劫·初雷」，附带雷霆特效；累计充值 ¥500,000 解锁渡劫一层，赠渡劫背景（权益待定，可随时调整）',
      '灵石积分 +25600，用于兑换渡劫护符',
      '头像框「雷劫之环」解锁',
      '服饰「劫云战袍」上线',
      '法宝「九天雷鼓」虚拟特效',
      '聊天气泡「仙雷洗礼」',
      '消费积分加成提升至 +45%',
      '动作「雷霆御身」解锁',
      '灵宠蜕变为「碧灵狐·雷劫」',
      '专属特效「仙雷护体」'
    ],
    milestone: {
      type: '整店欢聚夜',
      summary: '整店欢聚夜包场券 + 渡劫雷劫动态场景',
      rights: [
        {
          rightId: 'right_realm_tribulation_fullhouse',
          quantity: 1,
          description: '渡劫整店欢聚夜 1 次'
        }
      ]
    }
  },
  {
    id: 'realm_ascension',
    name: '飞升期',
    shortName: '飞升',
    description: '飞升在即，脱离凡俗成就真仙，享无尽寿元与更高维力量。',
    increment: 51200,
    thresholds: [1000000, 1050000, 1100000, 1150000, 1200000, 1250000, 1300000, 1350000, 1400000, 1450000],
    discount: 0.9,
    virtualRewards: [
      '称号「飞升·初羽」，获得仙羽光效；累计充值 ¥1,000,000 解锁飞升一层，赠飞升背景（权益待定，可随时调整）',
      '灵石积分 +51200，可兑换仙界供品',
      '头像框「飞升圣环」解锁',
      '服饰「仙尊羽衣」上线',
      '法宝「九霄云辇」虚拟座驾',
      '聊天气泡「仙韵缭绕」',
      '消费积分加成提升至 +50%',
      '动作「飞升之姿」解锁',
      '灵宠圆满为「碧灵狐·仙灵」',
      '终极虚拟形象套装「仙尊降世」'
    ],
    milestone: {
      type: '至尊黑卡',
      summary: '至尊黑卡 + 终极仙尊虚拟套装',
      rights: [
        {
          rightId: 'right_realm_immortal_blackcard',
          quantity: 1,
          description: '飞升至尊黑卡'
        }
      ]
    }
  }
];

const membershipRights = [
  {
    _id: 'right_realm_qi_drink',
    name: '灵泉饮品券',
    description: '炼气圆满奖励，赠送店内特调饮品 1 杯',
    applyReservation: false,
    validDays: 60
  },
  {
    _id: 'right_realm_foundation_room',
    name: '筑基专属包房券',
    description: '非高峰时段私人包房 1 小时，需提前预约',
    applyReservation: true,
    applyTimeRanges: [
      { start: '12:00', end: '24:00' },
      { start: '00:00', end: '06:00' }
    ],
    validDays: 120,
    meta: { roomUsageCount: 1 }
  },
  {
    _id: 'right_realm_core_gift',
    name: '金丹定制礼盒',
    description: '法宝主题礼盒套装 1 份，线下领取',
    applyReservation: false,
    validDays: 365
  },
  {
    _id: 'right_realm_nascent_cigar',
    name: '元婴精品雪茄',
    description: '甄选雪茄 1 支，需到店核验身份后领取',
    applyReservation: false,
    validDays: 365
  },
  {
    _id: 'right_realm_divine_wine',
    name: '化神珍藏红酒',
    description: '珍藏红酒 1 瓶，可预约到店领取',
    applyReservation: false,
    validDays: 365
  },
  {
    _id: 'right_realm_void_badge',
    name: '炼虚尊贵标识包',
    description: '发放实体徽章与虚拟标识，永久生效',
    applyReservation: false
  },
  {
    _id: 'right_realm_unity_avatar',
    name: '合体专属形象套装',
    description: '钻石级专属头像、光环与服饰整套',
    applyReservation: false
  },
  {
    _id: 'right_realm_great_vehicle_banquet',
    name: '大乘贵宾宴请券',
    description: '年度贵宾宴席邀请 1 次，需提前预约',
    applyReservation: false,
    validDays: 365
  },
  {
    _id: 'right_realm_tribulation_fullhouse',
    name: '渡劫整店欢聚夜',
    description: '整店包场券 1 次，提前 7 日预约',
    applyReservation: false,
    validDays: 180
  },
  {
    _id: 'right_realm_immortal_blackcard',
    name: '飞升至尊黑卡',
    description: '终身尊享黑卡服务与专属顾问',
    applyReservation: false
  },
  {
    _id: 'right_daytime_room',
    name: '日间包房体验券',
    description: '免费预订一次日间非高峰包房',
    applyReservation: true,
    applyTimeRanges: [{ start: '12:00', end: '18:00' }],
    validDays: 90,
    meta: { roomUsageCount: 1 }
  },
  {
    _id: 'right_full_day_room',
    name: '全天包房体验券',
    description: '免费预订一次全天任意时段包房',
    applyReservation: true,
    applyTimeRanges: [{ start: '00:00', end: '24:00' }],
    validDays: 120,
    meta: { roomUsageCount: 1 }
  },
  {
    _id: 'right_full_house',
    name: '整店包场券',
    description: '尊享整店包场特权一次，需提前预约',
    applyReservation: false,
    validDays: 180
  },
  {
    _id: 'right_premium_gift',
    name: '精品礼盒',
    description: '10 支雪茄 BHK56 礼盒',
    applyReservation: false,
    validDays: 365
  }
];

module.exports = {
  EXPERIENCE_PER_YUAN,
  COLLECTIONS,
  EXCLUDED_TRANSACTION_STATUSES,
  DEFAULT_ADMIN_ROLES,
  subLevelLabels,
  realmConfigs,
  membershipRights
};
